### PR TITLE
PLT-135 Showing "(Edited)" indicator if a message has been edited.

### PR DESCRIPTION
**Pull request by debanshuk**
#### Summary
Showing "(Edited)" indicator if a message has been edited.
I have done translation entry for only English.

#### Ticket Link
JIRA Ticket: https://mattermost.atlassian.net/browse/PLT-135
Github ticket: https://github.com/mattermost/platform/issues/4751

#### Checklist
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates

#### Screenshots
Attached are two screenshots showing the edited text in different types of messages with text formatting enabled and disabled respectively.
mm-1
mm-2

### CODE REVIEW

**Comment by enahum** [comment]
Hi @debanshuk

Can you please remove every translation and only leave the english one, as the other translations have to follow the translation process.

I'll take this oportunity to invite you to watch our [brown bag on localization](https://www.youtube.com/watch?v=cVRmzXjpp7Y) if you want to understand the process.

Thanks

**Comment by it33** [comment]
Tested: 
- basic functionality
- case with line wrapping
- case with 4000 characters edited to 3999

Looks good. 

There's at least one corner case I found during testing and have opened other tickets to address: 

- https://github.com/mattermost/platform/issues/4765

**Comment by debanshuk** [comment]
@enahum, Removed every translation entry apart from English and updated pull request.

**Comment by jasonblais** [comment]
Thanks @debanshuk! Great work with the implementation!

Passed PM review, tested on Windows 10, Chrome + approval from @it33.

----

FYI - created tickets for issues related to editing and replying messages above 4000 characters
 -  https://mattermost.atlassian.net/browse/PLT-4943
 - https://mattermost.atlassian.net/browse/PLT-4976

They're not related to this particular PR, and hence don't block it from moving on to dev review.

File: webapp/components/post_view/components/post_message_view.jsx [hunk 1]
```diff
@@ -2,28 +2,37 @@
 // See License.txt for license information.
 
 import React from 'react';
+import {intlShape, injectIntl, defineMessages, FormattedMessage} from 'react-intl';
 
 import * as TextFormatting from 'utils/text_formatting.jsx';
 import * as Utils from 'utils/utils.jsx';
 
-export default class PostMessageView extends React.Component {
+const messages = defineMessages({
```
Comment (enahum) [inline webapp/components/post_view/components/post_message_view.jsx]: We should avoid using this and try and use `<FormattedMessage>`

File: webapp/components/post_view/components/post_message_view.jsx [hunk 2]
```diff
@@ -47,9 +56,30 @@ export default class PostMessageView extends React.Component {
...
         if (!this.props.enableFormatting) {
-            return <span>{this.props.message}</span>;
+            return (
+                <span>
+                    {this.props.post.message}
+                    {
+                        this.isPostEdited() ?
+                            <span className='edited'>
+                                &nbsp;
+                                <FormattedMessage
```
Comment (enahum) [inline webapp/components/post_view/components/post_message_view.jsx]: Which you are using here, so I really don't understand why using `injectIntl` and `defineMessages`

File: webapp/components/post_view/components/post_message_view.jsx [hunk 3]
```diff
@@ -64,8 +94,21 @@ export default class PostMessageView extends React.Component {
...
             <span
                 onClick={Utils.handleFormattedTextClick}
-                dangerouslySetInnerHTML={{__html: TextFormatting.formatText(this.props.message, options)}}
+                dangerouslySetInnerHTML={{
+                    __html: (
+                        TextFormatting.formatText(this.props.post.message, options) +
+                        (
+                            this.isPostEdited() ?
+                                '<span class="edited">' +
+                                formatMessage(messages.edited) +
```
Comment (enahum) [inline webapp/components/post_view/components/post_message_view.jsx]: again, try and avoid this here

File: webapp/components/post_view/components/post_message_view.jsx (hunk 2 above)
Comment (debanshuk) [inline webapp/components/post_view/components/post_message_view.jsx]: @enahum, If `this.props.enableFormatting` is true then the above JSX is rendered, but if it's false then the one at line 103 is rendered. And in the one at line 103, we are passing formatted message (formatted using `TextFormatting.formatText`) to `dangerouslySetInnerHTML` property and there I am using `formatMessage` (and using `injectIntl` and `defineMessages`) as I needed to append a HTML string to the HTML string returned by `TextFormatting.formatText`.

Comment (debanshuk) [inline webapp/components/post_view/components/post_message_view.jsx]: @enahum, Or I can use something [html-to-react](https://www.npmjs.com/package/html-to-react) to remove use of `dangerouslySetInnerHTML` and then would also be able to remove use of `injectIntl`, `defineMessages` and ``formatMessage` and would be able to use `<FormattedMessage>`. Shall I do that?

Comment (enahum) [inline webapp/components/post_view/components/post_message_view.jsx]: you could use `Utils.localizeMessage('post_message_view.edited', '(edited)')` and get rid of the `injectIntl` and `defineMessages` as it does not need to evaluate any vars

**Comment by debanshuk** [comment]
@enahum, Update PR according to code review comments. Following are the changes:
* Using `Utils.localizeMessage`, instead of `formatMessage` and have gotten rid of `injectIntl` and `defineMessages`.
* Created a PostUtil called `isEdited(post)` (though this was not in review comments).

Though for reason CI build is failing.

**Comment by debanshuk** [comment]
Rebased branch from latest master, but CI build is still failing.

**Comment by debanshuk** [comment]
Looks like the CI build have now passed.

File: webapp/components/post_view/components/post_message_view.jsx [hunk 4]
```diff
@@ -64,7 +85,18 @@ export default class PostMessageView extends React.Component {
         return (
             <span
                 onClick={Utils.handleFormattedTextClick}
-                dangerouslySetInnerHTML={{__html: TextFormatting.formatText(this.props.message, options)}}
+                dangerouslySetInnerHTML={{
```
Comment (hmhealey) [inline webapp/components/post_view/components/post_message_view.jsx]: Can you change this not to use dangerouslySetInnerHTML? We need that for markdown and text formatting, but we shouldn't need it for the (edited) label. You can probably  just wrap this whole thing in a div if necessary like
```
return (
    <div>
        <span dangerouslySetInnerHTML=.../>
        <span className='edited'>...</span>
    </div>
);
````

Comment (debanshuk) [inline webapp/components/post_view/components/post_message_view.jsx]: @hmhealey, Done. Also, did a missing change in .post__body's style for compact view.

**Comment by debanshuk** [comment]
@enahum @hmhealey, All review changes has been done. Please re-review the update PR.

**Comment by esethna** [comment]
Hi @debanshuk, you'll notice that we had to revert this PR in https://github.com/mattermost/platform/pull/4854.

After it was merged and tested on our [nightly build server](https://pre-release.mattermost.com/core/channels/tickets) we noticed a number of bugs that were missed in the initial review, including:

1. https://mattermost.atlassian.net/browse/PLT-5052
2. https://mattermost.atlassian.net/browse/PLT-5055
3. https://mattermost.atlassian.net/browse/PLT-5056
4. Placement of the `(Edited)` text does not work in all cases.

Thanks again for your contribution, it is much appreciated. After the holidays we are going to re-review the design, then we'd be happy to loop you in if you're interested in working on this again and addressing the bugs we found?

**Comment by debanshuk** [comment]
@esethna, Sure, I would like to work for this again addressing bugs. 

I can see the reason of PLT-5052 is related to stye changes done by me regarding display types of `<p>` tags and there is incorrectly coded case (for first p tag) which would cause the bug.

Other issue ie, PLT-5055 and PLT-5056 are cases which I wasn't aware of and missed completely. I can work on them too.

**Comment by esethna** [comment]
Thanks @debanshuk! Let's hold off on working on this until the UX team has had a chance to discuss some of the issues, then we can circle back with you on the proposed approach.

**Comment by esethna** [comment]
@debanshuk thanks for your interest in working on the bug fixes for this feature. UX team has discussed and we are going to go with the original placement (`(edited)` appears at the end of the post text ) with one minor tweak being decreasing the opacity of the `(edited)` indicator so it's not as prominent. 

Please let me know if you are still intersted in re-submitting a PR for the feature addressing the bugs and UX concern?

**Comment by debanshuk** [comment]
@esethna, Yes I am interested in re-submitting a PR for the feature addressing the bugs and UX concern. Should I submit the new PR for the same Github issue?

**Comment by esethna** [comment]
@debanshuk yes please, that would excellent, thanks!

**Pull request merged**